File: testdata/sample-code/go/errors/custom_errors.go
package errors

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"runtime"
	"strings"
	"time"
)

// ============================================================================
// Base Error Types
// ============================================================================

// ErrorCode represents a specific error type
type ErrorCode string

const (
	// Validation errors (1xxx)
	ErrValidationFailed ErrorCode = "VALIDATION_FAILED"
	ErrRequiredField    ErrorCode = "REQUIRED_FIELD"
	ErrInvalidFormat    ErrorCode = "INVALID_FORMAT"
	ErrInvalidValue     ErrorCode = "INVALID_VALUE"
	ErrOutOfRange       ErrorCode = "OUT_OF_RANGE"

	// Authentication errors (2xxx)
	ErrUnauthorized       ErrorCode = "UNAUTHORIZED"
	ErrInvalidCredentials ErrorCode = "INVALID_CREDENTIALS"
	ErrTokenExpired       ErrorCode = "TOKEN_EXPIRED"
	ErrTokenInvalid       ErrorCode = "TOKEN_INVALID"
	ErrInsufficientScope  ErrorCode = "INSUFFICIENT_SCOPE"

	// Resource errors (3xxx)
	ErrNotFound       ErrorCode = "NOT_FOUND"
	ErrAlreadyExists  ErrorCode = "ALREADY_EXISTS"
	ErrConflict       ErrorCode = "CONFLICT"
	ErrResourceLocked ErrorCode = "RESOURCE_LOCKED"

	// Business logic errors (4xxx)
	ErrInsufficientFunds ErrorCode = "INSUFFICIENT_FUNDS"
	ErrLimitExceeded     ErrorCode = "LIMIT_EXCEEDED"
	ErrInvalidOperation  ErrorCode = "INVALID_OPERATION"
	ErrDuplicateEntry    ErrorCode = "DUPLICATE_ENTRY"

	// System errors (5xxx)
	ErrInternal           ErrorCode = "INTERNAL_ERROR"
	ErrDatabase           ErrorCode = "DATABASE_ERROR"
	ErrNetwork            ErrorCode = "NETWORK_ERROR"
	ErrTimeout            ErrorCode = "TIMEOUT"
	ErrServiceUnavailable ErrorCode = "SERVICE_UNAVAILABLE"
)

// Severity level for errors
type Severity string

const (
	SeverityDebug    Severity = "DEBUG"
	SeverityInfo     Severity = "INFO"
	SeverityWarning  Severity = "WARNING"
	SeverityError    Severity = "ERROR"
	SeverityCritical Severity = "CRITICAL"
)

// ============================================================================
// DomainError - Base custom error
// ============================================================================

// DomainError represents a domain-specific error with context
type DomainError struct {
	Code      ErrorCode   `json:"code"`
	Message   string      `json:"message"`
	Details   interface{} `json:"details,omitempty"`
	Severity  Severity    `json:"severity"`
	Timestamp time.Time   `json:"timestamp"`
	Stack     []string    `json:"-"`
	cause     error
}

func NewDomainError(code ErrorCode, message string) *DomainError {
	return &DomainError{
		Code:      code,
		Message:   message,
		Severity:  SeverityError,
		Timestamp: time.Now(),
		Stack:     captureStack(2),
	}
}

func (e *DomainError) Error() string {
	if e.cause != nil {
		return fmt.Sprintf("[%s] %s: %v", e.Code, e.Message, e.cause)
	}
	return fmt.Sprintf("[%s] %s", e.Code, e.Message)
}

func (e *DomainError) Unwrap() error {
	return e.cause
}

func (e *DomainError) WithCause(err error) *DomainError {
	e.cause = err
	return e
}

func (e *DomainError) WithDetails(details interface{}) *DomainError {
	e.Details = details
	return e
}

func (e *DomainError) WithSeverity(severity Severity) *DomainError {
	e.Severity = severity
	return e
}

// Is implements error matching for DomainError
func (e *DomainError) Is(target error) bool {
	t, ok := target.(*DomainError)
	if !ok {
		return false
	}
	return e.Code == t.Code
}

// ============================================================================
// Validation Errors
// ============================================================================

// ValidationError represents validation failures
type ValidationError struct {
	*DomainError
	Field string      `json:"field"`
	Value interface{} `json:"value,omitempty"`
	Rule  string      `json:"rule,omitempty"`
}

func NewValidationError(field, message string) *ValidationError {
	return &ValidationError{
		DomainError: NewDomainError(ErrValidationFailed, message),
		Field:       field,
	}
}

func NewRequiredFieldError(field string) *ValidationError {
	return &ValidationError{
		DomainError: NewDomainError(ErrRequiredField, fmt.Sprintf("%s is required", field)),
		Field:       field,
		Rule:        "required",
	}
}

func NewInvalidFormatError(field, format string) *ValidationError {
	return &ValidationError{
		DomainError: NewDomainError(ErrInvalidFormat, fmt.Sprintf("%s must be in %s format", field, format)),
		Field:       field,
		Rule:        "format",
		Value:       format,
	}
}

func NewOutOfRangeError(field string, min, max interface{}) *ValidationError {
	return &ValidationError{
		DomainError: NewDomainError(ErrOutOfRange, fmt.Sprintf("%s must be between %v and %v", field, min, max)),
		Field:       field,
		Rule:        "range",
		Details: map[string]interface{}{
			"min": min,
			"max": max,
		},
	}
}

// ValidationErrors aggregates multiple validation errors
type ValidationErrors struct {
	Errors []*ValidationError `json:"errors"`
}

func NewValidationErrors() *ValidationErrors {
	return &ValidationErrors{
		Errors: make([]*ValidationError, 0),
	}
}

func (v *ValidationErrors) Add(err *ValidationError) {
	v.Errors = append(v.Errors, err)
}

func (v *ValidationErrors) Error() string {
	if len(v.Errors) == 0 {
		return "validation failed"
	}

	var sb strings.Builder
	sb.WriteString("validation failed: ")
	for i, err := range v.Errors {
		if i > 0 {
			sb.WriteString(", ")
		}
		sb.WriteString(err.Message)
	}
	return sb.String()
}

func (v *ValidationErrors) HasErrors() bool {
	return len(v.Errors) > 0
}

// ============================================================================
// Authentication Errors
// ============================================================================

// AuthError represents authentication/authorization errors
type AuthError struct {
	*DomainError
	UserID   string `json:"user_id,omitempty"`
	Resource string `json:"resource,omitempty"`
	Required string `json:"required,omitempty"`
}

func NewUnauthorizedError(message string) *AuthError {
	return &AuthError{
		DomainError: NewDomainError(ErrUnauthorized, message),
	}
}

func NewInvalidCredentialsError() *AuthError {
	return &AuthError{
		DomainError: NewDomainError(ErrInvalidCredentials, "invalid email or password"),
	}
}

func NewTokenExpiredError() *AuthError {
	return &AuthError{
		DomainError: NewDomainError(ErrTokenExpired, "token has expired"),
	}
}

func NewInsufficientScopeError(required, actual string) *AuthError {
	return &AuthError{
		DomainError: NewDomainError(ErrInsufficientScope, "insufficient permissions"),
		Required:    required,
		Details: map[string]string{
			"required": required,
			"actual":   actual,
		},
	}
}

// ============================================================================
// Resource Errors
// ============================================================================

// NotFoundError represents resource not found
type NotFoundError struct {
	*DomainError
	ResourceType string `json:"resource_type"`
	ResourceID   string `json:"resource_id"`
}

func NewNotFoundError(resourceType, id string) *NotFoundError {
	return &NotFoundError{
		DomainError:  NewDomainError(ErrNotFound, fmt.Sprintf("%s not found: %s", resourceType, id)),
		ResourceType: resourceType,
		ResourceID:   id,
	}
}

func (e *NotFoundError) IsNotFound() bool {
	return true
}

// ConflictError represents resource conflict
type ConflictError struct {
	*DomainError
	ResourceType string `json:"resource_type"`
	ResourceID   string `json:"resource_id"`
}

func NewConflictError(resourceType, id, message string) *ConflictError {
	return &ConflictError{
		DomainError:  NewDomainError(ErrConflict, message),
		ResourceType: resourceType,
		ResourceID:   id,
	}
}

func NewAlreadyExistsError(resourceType, field, value string) *ConflictError {
	return &ConflictError{
		DomainError:  NewDomainError(ErrAlreadyExists, fmt.Sprintf("%s with %s '%s' already exists", resourceType, field, value)),
		ResourceType: resourceType,
		Details: map[string]string{
			"field": field,
			"value": value,
		},
	}
}

// ============================================================================
// Business Logic Errors
// ============================================================================

// BusinessError represents business rule violations
type BusinessError struct {
	*DomainError
	Rule    string                 `json:"rule"`
	Context map[string]interface{} `json:"context"`
}

func NewBusinessError(rule, message string) *BusinessError {
	return &BusinessError{
		DomainError: NewDomainError(ErrInvalidOperation, message),
		Rule:        rule,
		Context:     make(map[string]interface{}),
	}
}

func (e *BusinessError) WithContext(key string, value interface{}) *BusinessError {
	e.Context[key] = value
	return e
}

// InsufficientFundsError for financial operations
type InsufficientFundsError struct {
	*BusinessError
	AccountID string  `json:"account_id"`
	Balance   float64 `json:"balance"`
	Required  float64 `json:"required"`
	Shortfall float64 `json:"shortfall"`
}

func NewInsufficientFundsError(accountID string, balance, required float64) *InsufficientFundsError {
	shortfall := required - balance
	return &InsufficientFundsError{
		BusinessError: NewBusinessError("insufficient_funds", "insufficient funds for transaction"),
		AccountID:     accountID,
		Balance:       balance,
		Required:      required,
		Shortfall:     shortfall,
	}
}

// LimitExceededError for rate limits and quotas
type LimitExceededError struct {
	*BusinessError
	LimitType string    `json:"limit_type"`
	Limit     int64     `json:"limit"`
	Current   int64     `json:"current"`
	ResetAt   time.Time `json:"reset_at"`
}

func NewLimitExceededError(limitType string, limit, current int64, resetAt time.Time) *LimitExceededError {
	return &LimitExceededError{
		BusinessError: NewBusinessError("limit_exceeded", fmt.Sprintf("%s limit exceeded", limitType)),
		LimitType:     limitType,
		Limit:         limit,
		Current:       current,
		ResetAt:       resetAt,
	}
}

// ============================================================================
// System Errors
// ============================================================================

// SystemError represents internal system errors
type SystemError struct {
	*DomainError
	Component string `json:"component"`
	Operation string `json:"operation"`
}

func NewInternalError(message string) *SystemError {
	return &SystemError{
		DomainError: NewDomainError(ErrInternal, message),
	}
}

func NewDatabaseError(operation string, err error) *SystemError {
	return &SystemError{
		DomainError: NewDomainError(ErrDatabase, fmt.Sprintf("database error during %s", operation)).WithCause(err),
		Component:   "database",
		Operation:   operation,
	}
}

func NewNetworkError(operation string, err error) *SystemError {
	return &SystemError{
		DomainError: NewDomainError(ErrNetwork, fmt.Sprintf("network error during %s", operation)).WithCause(err),
		Component:   "network",
		Operation:   operation,
	}
}

func NewTimeoutError(operation string, timeout time.Duration) *SystemError {
	return &SystemError{
		DomainError: NewDomainError(ErrTimeout, fmt.Sprintf("timeout after %v during %s", timeout, operation)),
		Component:   "timeout",
		Operation:   operation,
	}
}

// ============================================================================
// HTTP Error Mapping
// ============================================================================

// HTTPError represents an error with HTTP status code
type HTTPError struct {
	error
	StatusCode int `json:"-"`
}

func (e HTTPError) Error() string {
	return e.error.Error()
}

func (e HTTPError) Status() int {
	return e.StatusCode
}

// MapErrorToHTTP converts domain errors to HTTP responses
func MapErrorToHTTP(err error) (int, interface{}) {
	var statusCode int
	var response map[string]interface{}

	switch e := err.(type) {
	case *ValidationError:
		statusCode = http.StatusBadRequest
		response = map[string]interface{}{
			"code":    e.Code,
			"message": e.Message,
			"field":   e.Field,
		}

	case *ValidationErrors:
		statusCode = http.StatusBadRequest
		response = map[string]interface{}{
			"code":    ErrValidationFailed,
			"message": e.Error(),
			"errors":  e.Errors,
		}

	case *AuthError:
		switch e.Code {
		case ErrUnauthorized, ErrInvalidCredentials:
			statusCode = http.StatusUnauthorized
		case ErrInsufficientScope:
			statusCode = http.StatusForbidden
		default:
			statusCode = http.StatusUnauthorized
		}
		response = map[string]interface{}{
			"code":    e.Code,
			"message": e.Message,
		}

	case *NotFoundError:
		statusCode = http.StatusNotFound
		response = map[string]interface{}{
			"code":    e.Code,
			"message": e.Message,
		}

	case *ConflictError:
		statusCode = http.StatusConflict
		response = map[string]interface{}{
			"code":    e.Code,
			"message": e.Message,
		}

	case *BusinessError:
		statusCode = http.StatusUnprocessableEntity
		response = map[string]interface{}{
			"code":    e.Code,
			"message": e.Message,
			"rule":    e.Rule,
		}

	case *InsufficientFundsError:
		statusCode = http.StatusPaymentRequired
		response = map[string]interface{}{
			"code":      e.Code,
			"message":   e.Message,
			"balance":   e.Balance,
			"required":  e.Required,
			"shortfall": e.Shortfall,
		}

	case *LimitExceededError:
		statusCode = http.StatusTooManyRequests
		response = map[string]interface{}{
			"code":     e.Code,
			"message":  e.Message,
			"limit":    e.Limit,
			"current":  e.Current,
			"reset_at": e.ResetAt,
		}

	case *SystemError:
		statusCode = http.StatusInternalServerError
		response = map[string]interface{}{
			"code":    e.Code,
			"message": "an internal error occurred",
		}

	default:
		statusCode = http.StatusInternalServerError
		response = map[string]interface{}{
			"code":    ErrInternal,
			"message": "an unexpected error occurred",
		}
	}

	// Add request ID if available (would come from context)
	if requestID := getRequestID(); requestID != "" {
		response["request_id"] = requestID
	}

	return statusCode, response
}

// Helper function (would be implemented in actual code)
func getRequestID() string {
	return "" // Placeholder
}

// ============================================================================
// Error Wrapping Utilities
// ============================================================================

// WrapError wraps an error with additional context
func WrapError(err error, message string) error {
	if err == nil {
		return nil
	}
	return fmt.Errorf("%s: %w", message, err)
}

// WrapErrorWithCode wraps an error with a domain error
func WrapErrorWithCode(err error, code ErrorCode, message string) *DomainError {
	if err == nil {
		return nil
	}

	domainErr := NewDomainError(code, message)
	domainErr.cause = err
	return domainErr
}

// IsNotFound checks if an error is a not found error
func IsNotFound(err error) bool {
	var nf *NotFoundError
	return errors.As(err, &nf)
}

// IsValidation checks if an error is a validation error
func IsValidation(err error) bool {
	var ve *ValidationError
	return errors.As(err, &ve)
}

// IsAuth checks if an error is an authentication error
func IsAuth(err error) bool {
	var ae *AuthError
	return errors.As(err, &ae)
}

// IsConflict checks if an error is a conflict error
func IsConflict(err error) bool {
	var ce *ConflictError
	return errors.As(err, &ce)
}

// ============================================================================
// Stack Trace Capture
// ============================================================================

func captureStack(skip int) []string {
	var stack []string
	for i := skip; i < skip+10; i++ {
		pc, file, line, ok := runtime.Caller(i)
		if !ok {
			break
		}
		fn := runtime.FuncForPC(pc)
		stack = append(stack, fmt.Sprintf("%s:%d %s", file, line, fn.Name()))
	}
	return stack
}

// ============================================================================
// Error Handling Middleware
// ============================================================================

// ErrorHandler middleware for HTTP handlers
type ErrorHandler struct {
	logger interface {
		Error(msg string, fields ...interface{})
	}
}

func NewErrorHandler(logger interface {
	Error(msg string, fields ...interface{})
}) *ErrorHandler {
	return &ErrorHandler{logger: logger}
}

func (h *ErrorHandler) Handle(next func(http.ResponseWriter, *http.Request) error) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if err := next(w, r); err != nil {
			h.handleError(w, r, err)
		}
	}
}

func (h *ErrorHandler) handleError(w http.ResponseWriter, r *http.Request, err error) {
	statusCode, response := MapErrorToHTTP(err)

	// Log error
	h.logger.Error("request failed",
		"error", err.Error(),
		"status", statusCode,
		"path", r.URL.Path,
		"method", r.Method,
	)

	// Send response
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(statusCode)
	json.NewEncoder(w).Encode(response)
}

// ============================================================================
// Retryable Error Detection
// ============================================================================

// IsRetryable determines if an error can be retried
func IsRetryable(err error) bool {
	var domainErr *DomainError
	if errors.As(err, &domainErr) {
		switch domainErr.Code {
		case ErrTimeout, ErrNetwork, ErrServiceUnavailable:
			return true
		case ErrInternal:
			// Some internal errors might be retryable
			return true
		}
	}

	// Network errors are retryable
	var netErr interface {
		Timeout() bool
		Temporary() bool
	}
	if errors.As(err, &netErr) && (netErr.Timeout() || netErr.Temporary()) {
		return true
	}

	return false
}

// ============================================================================
// Error Aggregation
// ============================================================================

// AggregateError collects multiple errors
type AggregateError struct {
	Errors []error `json:"errors"`
}

func NewAggregateError() *AggregateError {
	return &AggregateError{
		Errors: make([]error, 0),
	}
}

func (a *AggregateError) Add(err error) {
	if err != nil {
		a.Errors = append(a.Errors, err)
	}
}

func (a *AggregateError) Error() string {
	if len(a.Errors) == 0 {
		return "no errors"
	}

	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("%d errors occurred", len(a.Errors)))
	for i, err := range a.Errors {
		sb.WriteString(fmt.Sprintf("\n  %d: %v", i+1, err))
	}
	return sb.String()
}

func (a *AggregateError) HasErrors() bool {
	return len(a.Errors) > 0
}

// ============================================================================
// Error Factories
// ============================================================================

// Common error factories
var (
	ErrEmailRequired    = NewRequiredFieldError("email")
	ErrPasswordRequired = NewRequiredFieldError("password")
	ErrNameRequired     = NewRequiredFieldError("name")

	ErrInvalidEmail = NewInvalidFormatError("email", "email")
	ErrInvalidPhone = NewInvalidFormatError("phone", "E.164")
	ErrInvalidURL   = NewInvalidFormatError("url", "URL")

	ErrUserNotFound    = NewNotFoundError("user", "unknown")
	ErrProductNotFound = NewNotFoundError("product", "unknown")
	ErrOrderNotFound   = NewNotFoundError("order", "unknown")

	ErrEmailAlreadyExists = NewAlreadyExistsError("user", "email", "")
)

// ============================================================================
// Error Examples
// ============================================================================

func ExampleUsage() error {
	// Basic domain error
	err := NewDomainError(ErrValidationFailed, "invalid input")

	// With cause
	cause := errors.New("underlying error")
	err = NewDomainError(ErrDatabase, "query failed").WithCause(cause)

	// Validation error
	err = NewValidationError("email", "invalid email format")

	// Multiple validation errors
	ve := NewValidationErrors()
	ve.Add(NewRequiredFieldError("name"))
	ve.Add(NewInvalidFormatError("email", "email"))

	// Business error with context
	err = NewBusinessError("max_amount", "amount exceeds limit").
		WithContext("max_allowed", 1000).
		WithContext("requested", 1500)

	// Check error types
	switch e := err.(type) {
	case *NotFoundError:
		fmt.Printf("resource not found: %s/%s", e.ResourceType, e.ResourceID)
	case *ValidationError:
		fmt.Printf("validation failed for field %s: %s", e.Field, e.Message)
	case *AuthError:
		fmt.Printf("auth error: %s", e.Message)
	}

	// Check with errors.Is
	if errors.Is(err, ErrUserNotFound) {
		// Handle not found
	}

	return err
}

// ============================================================================
// JSON Marshaling
// ============================================================================

func (e *DomainError) MarshalJSON() ([]byte, error) {
	type Alias DomainError
	return json.Marshal(&struct {
		*Alias
		Error string `json:"error"`
	}{
		Alias: (*Alias)(e),
		Error: e.Error(),
	})
}

// ============================================================================
// Error Stringer
// ============================================================================

func (c ErrorCode) String() string {
	return string(c)
}

func (s Severity) String() string {
	return string(s)
}

File: testdata/sample-code/nodejs/errors/custom-errors.js
/**
 * Custom Error Classes
 * Demonstrates different patterns for creating and using custom error types
 */

// ============================================================================
// Base Custom Error
// ============================================================================

/**
 * Base application error class
 * Extends the native Error with additional properties
 */
class AppError extends Error {
    constructor(message, statusCode = 500, code = 'INTERNAL_ERROR') {
        super(message);
        this.name = this.constructor.name;
        this.statusCode = statusCode;
        this.code = code;
        this.isOperational = true;
        this.timestamp = new Date().toISOString();
        
        // Capture stack trace
        Error.captureStackTrace(this, this.constructor);
    }
    
    toJSON() {
        return {
            name: this.name,
            message: this.message,
            code: this.code,
            statusCode: this.statusCode,
            timestamp: this.timestamp,
            stack: this.stack
        };
    }
}

// ============================================================================
// HTTP Status Based Errors
// ============================================================================

/**
 * 400 Bad Request Error
 */
class BadRequestError extends AppError {
    constructor(message = 'Bad request', code = 'BAD_REQUEST') {
        super(message, 400, code);
    }
}

/**
 * 401 Unauthorized Error
 */
class UnauthorizedError extends AppError {
    constructor(message = 'Unauthorized', code = 'UNAUTHORIZED') {
        super(message, 401, code);
    }
}

/**
 * 403 Forbidden Error
 */
class ForbiddenError extends AppError {
    constructor(message = 'Forbidden', code = 'FORBIDDEN') {
        super(message, 403, code);
    }
}

/**
 * 404 Not Found Error
 */
class NotFoundError extends AppError {
    constructor(resource = 'Resource', id = null) {
        const message = id 
            ? `${resource} with id ${id} not found`
            : `${resource} not found`;
        super(message, 404, 'NOT_FOUND');
        this.resource = resource;
        this.resourceId = id;
    }
}

/**
 * 409 Conflict Error
 */
class ConflictError extends AppError {
    constructor(message = 'Conflict occurred', code = 'CONFLICT') {
        super(message, 409, code);
    }
}

/**
 * 422 Unprocessable Entity Error
 */
class ValidationError extends AppError {
    constructor(message = 'Validation failed', errors = []) {
        super(message, 422, 'VALIDATION_ERROR');
        this.errors = errors;
    }
    
    addError(field, message) {
        this.errors.push({ field, message });
    }
}

/**
 * 429 Too Many Requests Error
 */
class RateLimitError extends AppError {
    constructor(message = 'Too many requests', retryAfter = 60) {
        super(message, 429, 'RATE_LIMIT_EXCEEDED');
        this.retryAfter = retryAfter;
    }
}

// ============================================================================
// Domain-Specific Errors
// ============================================================================

/**
 * User-related errors
 */
class UserNotFoundError extends NotFoundError {
    constructor(userId) {
        super('User', userId);
        this.code = 'USER_NOT_FOUND';
    }
}

class UserAlreadyExistsError extends ConflictError {
    constructor(email) {
        super(`User with email ${email} already exists`, 'USER_ALREADY_EXISTS');
        this.email = email;
    }
}

class InvalidCredentialsError extends UnauthorizedError {
    constructor() {
        super('Invalid email or password', 'INVALID_CREDENTIALS');
    }
}

class AccountLockedError extends ForbiddenError {
    constructor(userId, lockedUntil) {
        super(`Account locked until ${lockedUntil}`, 'ACCOUNT_LOCKED');
        this.userId = userId;
        this.lockedUntil = lockedUntil;
    }
}

/**
 * Product-related errors
 */
class ProductNotFoundError extends NotFoundError {
    constructor(productId) {
        super('Product', productId);
        this.code = 'PRODUCT_NOT_FOUND';
    }
}

class InsufficientStockError extends BadRequestError {
    constructor(productId, requested, available) {
        super(
            `Insufficient stock for product ${productId}. Requested: ${requested}, Available: ${available}`,
            'INSUFFICIENT_STOCK'
        );
        this.productId = productId;
        this.requested = requested;
        this.available = available;
    }
}

class ProductOutOfStockError extends BadRequestError {
    constructor(productId) {
        super(`Product ${productId} is out of stock`, 'OUT_OF_STOCK');
        this.productId = productId;
    }
}

/**
 * Order-related errors
 */
class OrderNotFoundError extends NotFoundError {
    constructor(orderId) {
        super('Order', orderId);
        this.code = 'ORDER_NOT_FOUND';
    }
}

class OrderCannotBeCancelledError extends BadRequestError {
    constructor(orderId, status) {
        super(
            `Order ${orderId} cannot be cancelled in status: ${status}`,
            'ORDER_CANNOT_BE_CANCELLED'
        );
        this.orderId = orderId;
        this.status = status;
    }
}

class InvalidOrderStatusError extends BadRequestError {
    constructor(orderId, currentStatus, newStatus) {
        super(
            `Cannot transition order ${orderId} from ${currentStatus} to ${newStatus}`,
            'INVALID_ORDER_STATUS'
        );
        this.orderId = orderId;
        this.currentStatus = currentStatus;
        this.newStatus = newStatus;
    }
}

/**
 * Payment-related errors
 */
class PaymentFailedError extends AppError {
    constructor(message = 'Payment failed', paymentMethod = null) {
        super(message, 402, 'PAYMENT_FAILED');
        this.paymentMethod = paymentMethod;
    }
}

class InsufficientFundsError extends PaymentFailedError {
    constructor(required, available) {
        super(`Insufficient funds. Required: ${required}, Available: ${available}`);
        this.required = required;
        this.available = available;
        this.code = 'INSUFFICIENT_FUNDS';
    }
}

class CardDeclinedError extends PaymentFailedError {
        super('Card was declined', 'CREDIT_CARD');
        this.code = 'CARD_DECLINED';
    }
}

class InvalidPaymentMethodError extends BadRequestError {
    constructor(method) {
        super(`Invalid payment method: ${method}`, 'INVALID_PAYMENT_METHOD');
        this.method = method;
    }
}

// ============================================================================
// Database Errors
// ============================================================================

class DatabaseError extends AppError {
    constructor(message = 'Database error occurred', originalError = null) {
        super(message, 500, 'DATABASE_ERROR');
        this.originalError = originalError?.message;
    }
}

class ConnectionError extends DatabaseError {
    constructor(host, port) {
        super(`Failed to connect to database at ${host}:${port}`);
        this.code = 'DB_CONNECTION_ERROR';
        this.host = host;
        this.port = port;
    }
}

class QueryError extends DatabaseError {
    constructor(query, originalError) {
        super(`Query execution failed: ${originalError?.message}`);
        this.code = 'DB_QUERY_ERROR';
        this.query = query;
    }
}

class DuplicateKeyError extends DatabaseError {
    constructor(collection, key, value) {
        super(`Duplicate key error in ${collection}: ${key} = ${value}`);
        this.code = 'DUPLICATE_KEY';
        this.collection = collection;
        this.key = key;
        this.value = value;
    }
}

// ============================================================================
// External Service Errors
// ============================================================================

class ExternalServiceError extends AppError {
    constructor(service, message = 'External service error', statusCode = 503) {
        super(`${service}: ${message}`, statusCode, 'EXTERNAL_SERVICE_ERROR');
        this.service = service;
    }
}

class ServiceTimeoutError extends ExternalServiceError {
    constructor(service, timeout) {
        super(service, `Request timed out after ${timeout}ms`, 504);
        this.code = 'SERVICE_TIMEOUT';
        this.timeout = timeout;
    }
}

class ServiceUnavailableError extends ExternalServiceError {
    constructor(service) {
        super(service, 'Service is unavailable', 503);
        this.code = 'SERVICE_UNAVAILABLE';
    }
}

// ============================================================================
// File System Errors
// ============================================================================

class FileSystemError extends AppError {
    constructor(message = 'File system error', path = null) {
        super(message, 500, 'FILE_SYSTEM_ERROR');
        this.path = path;
    }
}

class FileNotFoundError extends FileSystemError {
    constructor(path) {
        super(`File not found: ${path}`, path);
        this.code = 'FILE_NOT_FOUND';
        this.statusCode = 404;
    }
}

class PermissionDeniedError extends FileSystemError {
    constructor(path) {
        super(`Permission denied: ${path}`, path);
        this.code = 'PERMISSION_DENIED';
        this.statusCode = 403;
    }
}

class FileExistsError extends FileSystemError {
    constructor(path) {
        super(`File already exists: ${path}`, path);
        this.code = 'FILE_EXISTS';
        this.statusCode = 409;
    }
}

// ============================================================================
// Configuration Errors
// ============================================================================

class ConfigurationError extends AppError {
    constructor(message = 'Configuration error') {
        super(message, 500, 'CONFIG_ERROR');
    }
}

class MissingConfigError extends ConfigurationError {
    constructor(key) {
        super(`Missing required configuration: ${key}`);
        this.code = 'MISSING_CONFIG';
        this.key = key;
    }
}

class InvalidConfigError extends ConfigurationError {
        super(`Invalid configuration value for: ${key}`);
        this.code = 'INVALID_CONFIG';
        this.key = key;
        this.value = value;
    }
}

// ============================================================================
// Export all error classes
// ============================================================================

module.exports = {
    // Base
    AppError,
    
    // HTTP status based
    BadRequestError,
    UnauthorizedError,
    ForbiddenError,
    NotFoundError,
    ConflictError,
    ValidationError,
    RateLimitError,
    
    // Domain specific - User
    UserNotFoundError,
    UserAlreadyExistsError,
    InvalidCredentialsError,
    AccountLockedError,
    
    // Domain specific - Product
    ProductNotFoundError,
    InsufficientStockError,
    ProductOutOfStockError,
    
    // Domain specific - Order
    OrderNotFoundError,
    OrderCannotBeCancelledError,
    InvalidOrderStatusError,
    
    // Domain specific - Payment
    PaymentFailedError,
    InsufficientFundsError,
    CardDeclinedError,
    InvalidPaymentMethodError,
    
    // Database
    DatabaseError,
    ConnectionError,
    QueryError,
    DuplicateKeyError,
    
    // External services
    ExternalServiceError,
    ServiceTimeoutError,
    ServiceUnavailableError,
    
    // File system
    FileSystemError,
    FileNotFoundError,
    PermissionDeniedError,
    FileExistsError,
    
    // Configuration
    ConfigurationError,
    MissingConfigError,
    InvalidConfigError
};